src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -176,4 +193,102 @@ private void serializeModel(TrainingJob trainingJob, ActionListener<IndexRespons
+            }
+        }
+    }
+
+    protected void updateModelsNewCluster() throws IOException, InterruptedException, ExecutionException {
+        if (modelDao.isCreated()) {
+            List<String> modelIds = searchModelIds();
+            for (String modelId : modelIds) {
+                Model model = modelDao.get(modelId);
+                ModelMetadata modelMetadata = model.getModelMetadata();
+                if (modelMetadata.getState().equals(ModelState.TRAINING)) {
+                    modelMetadata.setState(ModelState.FAILED);
+                    modelMetadata.setError("Training failed due to a cluster crash");
+                    modelDao.update(model, new ActionListener<IndexResponse>() {
+                        @Override
+                        public void onResponse(IndexResponse indexResponse) {
+                            logger.info("Model " + indexResponse.getId() + " updated successfully");
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.info("Failed to update model", e);
+                        }
+                    });
+                }
+            }
+        }
+    }
+
+    protected void updateModelsNodesRemoved(List<DiscoveryNode> removedNodes) throws IOException, InterruptedException, ExecutionException {
+        if (modelDao.isCreated()) {
+            List<String> modelIds = searchModelIds();
+            for (DiscoveryNode removedNode : removedNodes) {
+                for (String modelId : modelIds) {
+                    Model model = modelDao.get(modelId);
+                    ModelMetadata modelMetadata = model.getModelMetadata();
+                    if (modelMetadata.getNodeAssignment().equals(removedNode.getEphemeralId())
+                        && modelMetadata.getState().equals(ModelState.TRAINING)) {
+                        modelMetadata.setState(ModelState.ZOMBIE);
+                        modelMetadata.setError("A node dropped and left the model training process in a zombie state");
+                        modelDao.update(model, new ActionListener<IndexResponse>() {
+                            @Override
+                            public void onResponse(IndexResponse indexResponse) {
+                                logger.info("Model " + indexResponse.getId() + " updated successfully");
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                logger.info("Failed to update model", e);
+                            }
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    private List<String> searchModelIds() throws IOException, InterruptedException {
+        List<String> modelIds = new ArrayList<String>();
+        CountDownLatch latch = new CountDownLatch(1);
```

Comment:
@ryanbogan Can we use CompletableFuture here? This CountDownLatch is an old way to get async method calls. 
@navneet1v  any thoughts here?

Reply:
I tried using the completable future, but since this is running on a different thread already there was an error that crashed the cluster

Reply:
I see.